rust/src/mqtt/mqtt.rs

```diff
@@ -589,6 +624,64 @@ impl MQTTState {
         tx.tx_data.set_event(event as u8);
         self.transactions.push_back(tx);
     }
+
+    fn mqtt_frames(&mut self, flow: *const Flow, stream_slice: &StreamSlice, input: &MQTTMessage) {
+        let hdr = stream_slice.as_slice();
+        //MQTT payload has a fixed header of 2 bytes
+        let _mqtt_hdr = Frame::new(flow, stream_slice, hdr, 2, MQTTFrameType::Header as u8);
+        SCLogDebug!("mqtt_hdr Frame {:?}", _mqtt_hdr);
+        let rem_length = input.header.remaining_length as usize;
+        let data = &hdr[2..rem_length + 2];
+        let _mqtt_data = Frame::new(
+            flow,
+            stream_slice,
+            data,
+            rem_length as i64,
+            MQTTFrameType::Data as u8,
+        );
+        SCLogDebug!("mqtt_data Frame {:?}", _mqtt_data);
+    }
+    fn mqtt_frames_trunc(
+        &mut self, flow: *const Flow, stream_slice: &StreamSlice, input: &MQTTTruncatedData,
+        current: &[u8], msg: &MQTTMessage,
+    ) {
+        let hdr = stream_slice.as_slice();
+        let hdr_length = input.skipped_length - msg.header.remaining_length as usize;
+        let _mqtt_hdr = Frame::new(
+            flow,
+            stream_slice,
+            hdr,
+            hdr_length as i64,
+            MQTTFrameType::Header as u8,
+        );
+        SCLogDebug!("mqtt_hdr Frame {:?}", _mqtt_hdr);
+
+        if input.skipped_length >= current.len() {
+            //taking current.len() as reference as trunc.skipped_length >= current.len()
+
+            let rem_length = current.len() - hdr_length;
+            let data = &hdr[hdr_length..current.len()];
+            let _mqtt_data = Frame::new(
+                flow,
+                stream_slice,
+                data,
+                rem_length as i64,
+                MQTTFrameType::Data as u8,
+            );
```

Comment:
Following my previous comment, I think that if we really want to offer any insight into this truncated data, we would have to pass `rem` (https://github.com/OISF/suricata/blob/master/rust/src/mqtt/mqtt.rs#L457) to `mqtt_frames_trunc` (cf https://github.com/OISF/suricata/blob/master/rust/src/mqtt/parser.rs#L663)... 🤔

Reply:
This might work too. but I am passing the current buffer which also represents the full input buffer. I have based my length calculations on this.
Do you still want me to look into this?

Reply:
From running the tests locally, your approach seems to be working. So I think we can move on ;)

Reply:
Based on your comment on the test frames PR , I will make a new PR for frames to add trunc_data frames